### PR TITLE
Prevent unnecessary fleetworkspace RoleBinding updates

### PR DESCRIPTION
**Pull request by moio**
## Problem
The `restrictedAdminGlobalBindingsFleet` controller handler continuously rewrites `fleetworkspace` RoleBindings even in absence of any changes.

In many cases, that is not a problem, because:

1. it only fires every 5 minutes ([kubelet's nodeStatusReportFrequency](https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/) - because kubelet updates `v1.Node`, which triggers `nodesSyncer`, which Enqueues `v3.Node`, which fires `cluster-stats`, which Enqueues `v3.Cluster`, which has a related resource `restricted-admin-cluster` on `v3.GlobalRoleBindings` which ultimately results in `restrictedAdminGlobalBindingsFleet` firing
2. `fleetworkspace` RoleBindings exist only if Restricted Admin roles are used (2 per Restricted Admin)
3. the total number of  `fleetworkspace` RoleBindings grows linearly with the number of downstream clusters (they will not be many if clusters are few)

However:
1. if clusters are not connected/not available, the [cluster-deploy](https://github.com/rancher/rancher/blob/6e9989c790fe975cc3e9f8db5b894f3a2dc58e2f/pkg/controllers/management/clusterdeploy/clusterdeploy.go#L84) handler will start to fail and be executed more often. Every failure re-Enqueues `v3.Cluster` which indirectly fires `restrictedAdminGlobalBindingsFleet` (see above)
    - failure happens because it executes [clusterDeploy.getNodeAgentImage, which tries to list DaemonSets](https://github.com/rancher/rancher/blob/6e9989c790fe975cc3e9f8db5b894f3a2dc58e2f/pkg/controllers/management/clusterdeploy/clusterdeploy.go#L473-L479) from the disconnected downstream cluster
    - listing of DaemonSets will time out after ~30 seconds due to how TunnelServer's remotedialer mechanism works
    - re-Enqueuing will happen as fast as it can (in my setup every ~45s) for the first 30 attempts, then every 2 minutes from the 31th attempt on. These times are due to the [standard lasso requeuing mechanism](https://github.com/rancher/lasso/pull/50)
2. many accounts could be given the Restricted Admin role, increasing the number of executions
3. many downstream clusters may be registered, increasing the number of executions

In such cases a sensible amount of CPU time in Rancher and kubeapiserver can be consumed, along with increased WebSocket traffic for RoleBinding updates and generated garbage.
 
## Solution
Refactor `restrictedAdminGlobalBindingsFleet` so that it updates `RoleBindings` only when they were actually changed, noop otherwise.
 
## Testing
Reproducer in issue

### Manual Testing
Tested per reproducer

### Automated Testing
Unit tests added as well to cover the previously uncovered method and make sure `Update` is called only when necessary.

Summary: nothing else to do FMPOV

## QA Testing Considerations
No difference expected between new installs and upgrades.

### Regressions Considerations
In principle, Fleet integration could be affected.

Existing / newly added automated tests that provide evidence there are no regressions:
* existing Fleet integration tests should suffice.

### CODE REVIEW

File: pkg/controllers/management/restrictedadminrbac/fleet.go
Comment (MbolotSuse) [inline pkg/controllers/management/restrictedadminrbac/fleet.go:92]: Can we return nil here in the case that we got an already exists error above? I think that in the case that we have a binding which already exists we need to go through the below integrity check, right?

File: pkg/controllers/management/restrictedadminrbac/fleet.go
Comment (MbolotSuse) [inline pkg/controllers/management/restrictedadminrbac/fleet.go:97]: Nit: You might get some better performance out of pre-creating the binding that you want and then issuing an update right way if any field is determined as invalid.

I'm marking this as a nit since although several of these fields (subjects/ownerRefs) are slices of complex objects, in practice they will only contain only a few entries, and I think `reflect.DeepEqual` checks the length so I don't think this will end up having a significant performance impact.

File: pkg/controllers/management/restrictedadminrbac/fleet_test.go
Comment (MbolotSuse) [inline pkg/controllers/management/restrictedadminrbac/fleet_test.go:13]: I'd also recommend adding tests for the following cases: 
- We get an error other than not found on the get call, that gets returned up
- We get an error when creating the binding, that gets returned up
- The existing role has different ownerrefs, we update
- The existing role has different rolerefs, we update
- The existing role has different subjects, we update
- We tried updating but got an error, that is returned up

Lastly, generally speaking it's a good practice to test the handler itself (i.e. the `ensureRestricedAdminForFleet` function) since those are closer to "public" functions that have an input/output interface (rather than private functions that serve as implementation details). In this case, the public function is relatively simple, and we could get some additional coverage out of testing that directly. But since you didn't actually change it, I don't think it's required to get this PR in.

**Review by MbolotSuse — CHANGES_REQUESTED** [review]
Thank you for adding tests.

One overall comment: I'm not sure that I fully understand how this resolves the linked issue. While network calls will undoubtedly be more expensive than the deepequals that this PR uses, I would expect to see fairly similar CPU/memory spikes, since the same controllers are still being re-enqueued with this change. Though I may be underestimating just how bad the performance is for the Update calls in comparison to the DeepEquals checks.

File: pkg/controllers/management/restrictedadminrbac/fleet.go
Comment (rmweir) [reply pkg/controllers/management/restrictedadminrbac/fleet.go:97]: I agree that this is probably unnecessary. If it is kept, can the comparison logic be set in its own function? I think this would improve tests as well since the tests' main concern is that logic.

**Review by rmweir — CHANGES_REQUESTED** [review]

**Comment by moio** [comment]
> One overall comment: I'm not sure that I fully understand how this resolves the linked issue. While network calls will undoubtedly be more expensive than the deepequals that this PR uses, I would expect to see fairly similar CPU/memory spikes, since the same controllers are still being re-enqueued with this change. Though I may be underestimating just how bad the performance is for the Update calls in comparison to the DeepEquals checks.

There are several reasons why this approach is expected to have better performance:

 - as you noted, network access is avoided. Leaving network I/O aside for a moment, network access code actually represents most of the _CPU_ time (highlighted in pink in the detail flame graph below):

image

 - as a secondary consequence, structs do not have to be marshaled/unmarshaled in order to go over the network, and that is another relevant part of the CPU cost, as shown below:
image

 - additionally, all `kube-apiserver` load is avoided. We do not have a flame graph (and that is actually impossible as the particular user is on AKS), but it is probably safe to assume it'll be at least similar to the loads in the previous two points (network + unmarshaling/marshaling)
 - furthermore all `etcd` load is avoided, and crucially that includes network and disk I/O because a commit to disk (actually to the quorum) will be needed. The reason is that the **original code creates a brand new object every time the handler fires, instead of reading and then possibly updating an existing object**. That implies a new revision of the object is unconditionally created, which brings us to the last problem:
 - unnecessary watch traffic, and the load on watchers (lasso-based and browser-based, including Steve, in case something is observing rolebindings)
 
Does the above sound correct to you?

PS. in any case, I am preparing a debug image for this user and they will test it repeating the flame graph collection, which should provide a definitive test for the points above.

File: pkg/controllers/management/restrictedadminrbac/fleet.go
Comment (moio) [reply pkg/controllers/management/restrictedadminrbac/fleet.go:92]: That was the previous behavior, and from my perspective it's reasonable.

Yes, there is a small chance that between the Get call and the Create call somebody else has created a RoleBinding with the same name that also happens to be incorrect. If it's Rancher code, then that would be a bug that should be fixed rather than worked around here. If it's a user, I'd let that be reconciled on the next handler call rather than listing -> creating -> listing again -> potentially updating.

Does that make sense to you?

File: pkg/controllers/management/restrictedadminrbac/cluster_test.go
Comment (moio) [reply pkg/controllers/management/restrictedadminrbac/cluster_test.go:4]: Addressed as of 36fafbd7d

File: pkg/controllers/management/restrictedadminrbac/fleet.go
Comment (moio) [reply pkg/controllers/management/restrictedadminrbac/fleet.go:78]: I typically like it better to handle one condition at a time to keep from nesting, but I am fine either way in this case. Addressed as of 6a153b0bf

File: pkg/controllers/management/restrictedadminrbac/fleet.go
Comment (moio) [reply pkg/controllers/management/restrictedadminrbac/fleet.go:74]: That was the previous behavior, but I agree wrapping makes more sense here. Addressed as of 04b4a97dd

File: pkg/controllers/management/restrictedadminrbac/fleet.go
Comment (moio) [reply pkg/controllers/management/restrictedadminrbac/fleet.go:97]: Pre-creating a binding and passing it to `Update` has the disadvantage of potentially changing present and future fields I am not interested into, so my thinking was to avoid that. However, I do see your objection that once dirty is set to true, any further checks are wasted.

I also agree it will probably not make a big performance difference, still, in the spirit of trying to set a good example I reworked the logic to enable short-circuiting of checks.

Please let me know if it is acceptable as of e63a011a7.

File: pkg/controllers/management/restrictedadminrbac/fleet_test.go
Comment (moio) [reply pkg/controllers/management/restrictedadminrbac/fleet_test.go:13]: Suggested tests added as of 7fc80114d

File: pkg/controllers/management/restrictedadminrbac/fleet.go
Comment (MbolotSuse) [reply pkg/controllers/management/restrictedadminrbac/fleet.go:92]: > That was the previous behavior

I can resolve this on that basis, I don't think it makes sense to require fixing other issues as part of this PR.

That being said, I do not agree with the logic that this behavior is desired. The binding that this code is creating is managed by Rancher, IMO any corruption/change of that binding should be corrected as quickly as possible. Relying on the next handler call to fix it leaves an indeterminate time between when the corrupt binding was created and when it will be fixed (this could be called soon if a relatedResource change, or could wait a long time between now and the next adjustment).

File: pkg/controllers/management/restrictedadminrbac/fleet.go
Comment (MbolotSuse) [reply pkg/controllers/management/restrictedadminrbac/fleet.go]: Re-worked logic LGTM, I like the use of the or statement to effectively short-circuti.

**Comment by MbolotSuse** [comment]
@moio Thanks for the detailed explanation, that does make it appear that I was wrong about the overall impact of the update calls (it does end up being that much more expensive).

File: pkg/controllers/management/restrictedadminrbac/fleet_test.go
Comment (MbolotSuse) [inline pkg/controllers/management/restrictedadminrbac/fleet_test.go:190]: Here, and in the other places in the file, you are using `t` as the `*testing.T` arg. This is the test arg for the entire test function, so if you have a failure here it can only tell you that the entire test failed. For example, if I change this line to `assert.NotEqual(t, rb, expected)`, I see:

```
--- FAIL: Test_rbaccontroller_ensureRolebinding (0.00s)
    fleet_test.go:190: 
...
                Test:           Test_rbaccontroller_ensureRolebinding

```
This makes it hard to figure out exactly what sub-test failed. However, if you change `t` to `c.t` (which is the `testing.T` object for this sub-test, set on the mockController when you constructed that), you see:

```
--- FAIL: Test_rbaccontroller_ensureRolebinding (0.00s)
    --- FAIL: Test_rbaccontroller_ensureRolebinding/one_previously_existing_incorrect_rolebinding_(wrong_Subjects) (0.00s)
        fleet_test.go:190:
...
                Test:           Test_rbaccontroller_ensureRolebinding/one_previously_existing_incorrect_rolebinding_(wrong_Subjects)
```

Which tells you exactly what sub-test failed, making it much easier to track down failures.

In short:
- Change `t` to `c.t`, both here and in other places where this used.

File: pkg/controllers/management/restrictedadminrbac/fleet_test.go
Comment (MbolotSuse) [inline pkg/controllers/management/restrictedadminrbac/fleet_test.go:268]: Nit: You don't have to change this, but generally we like to have these tests run in parallel. The actual difference is not large, so you don't need to do it, but it's definitely there.
```suggestion
	for _, tt := range tests {
	        tt := tt
		t.Run(tt.name, func(t *testing.T) {
		        t.Parallel()
			mockCtrl := newMockController(t)
```

File: pkg/controllers/management/restrictedadminrbac/fleet.go
Comment (moio) [reply pkg/controllers/management/restrictedadminrbac/fleet.go:92]: I agree this is best settled in a separate PR.

Thanks also for the insights, I see where you are coming from.

> The binding that this code is creating is managed by Rancher, IMO any corruption/change of that binding should be corrected as quickly as possible.

Curiosity: would it be possible or even good to have similar "protected objects" actually completely guarded by Rancher's Webhook, so that nothing but Rancher can actually touch them? Such a design would make this whole "drift detection" logic useless (after, possibly, one last round of cleanup).

File: pkg/controllers/management/restrictedadminrbac/fleet_test.go
Comment (moio) [reply pkg/controllers/management/restrictedadminrbac/fleet_test.go:190]: Thanks for the mini-lesson, really appreciated the detail and clarity. Addressed as 83b064d2b

File: pkg/controllers/management/restrictedadminrbac/fleet_test.go
Comment (moio) [reply pkg/controllers/management/restrictedadminrbac/fleet_test.go]: Another small thing learned today, excellent! Addressed as 46985b92b

**Review by rmweir — APPROVED** [review]
LGTM

**Comment by moio** [comment]
User report from debug image with changes in this PR came back positive and confirmed theories above.

Before:

Screenshot 2023-10-03 at 13 44 41


After:
Screenshot 2023-10-03 at 13 44 48

@MbolotSuse can you please approve or let me know if anything else needs to be done before merging?

**Comment by MbolotSuse** [comment]
@moio given our current timelines (and goals with this PR), I'm going to hold off approving until the 2.8.0 release goes live to avoid an accidental merge and inclusion in 2.8.0.

**Comment by moio** [comment]
@MbolotSuse can I get your approval now that https://github.com/rancher/rancher/releases/tag/v2.8.0 is out?

Thanks in advance

**Review by MbolotSuse — APPROVED** [review]

**Comment by MbolotSuse** [comment]
@moio LGTM - please squash the commits when merging to keep the commit log clean.

**Pull request merged**
